js/general_patterns.md

```diff
@@ -0,0 +1,210 @@
+# General Patterns
+
+This document provides an overview of the general architecture and programming
+patterns that will be used throughout the JavaScript API. These patterns have
+been chosen because of the flexibility they provide in integrating with native
+APIs and ease to build new interfaces from, and for their ability to create
+performant and more optimizable code.
+
+
+
+### Asynchronous Calls
+
+The asynchronous callback pattern is very similar to `setTimeout()`. With the
+exception that required arguments must always go first. After the callback, N
+arguments can be passed that will propagate to the callback.
+
+Example of pattern:
+```js
+asyncCall(<arguments>, callback[, ...args])
+```
+
+`<arguments>` are the arguments required by the function call.
+
+`callback` is the function that is executed when the asynchronous operation is
+complete.
+
+`...args` are the additional values that will propagate to the callback.
```

Comment:
I decided to follow the callback with propagating arguments because it follows closely with the only other common API I see it used. With timers.

Okay. I'll add a paragraph stating all parameters are always required. This will be a great change on several levels.